openmeter/app/stripe/service/billing.go

```diff
@@ -0,0 +1,192 @@
+			return nil
+		}
+	}
+
+	var validationErrors *billing.InvoiceTriggerValidationInput
+	if input.GetValidationErrors != nil {
+		stripeValidationErrors, err := input.GetValidationErrors(stripeInvoice)
+		if err != nil {
+			s.logger.Error("failed to get validation errors", "error", err)
+			return err
+		}
+
+		if stripeValidationErrors != nil {
+			validationErrors = &billing.InvoiceTriggerValidationInput{
+				Operation: billing.InvoiceOpInitiatePayment,
+				Errors: lo.Map(stripeValidationErrors.Errors, func(stripeErr *stripe.Error, _ int) error {
+					return stripeErrorToValidationError(stripeErr)
+				}),
+			}
+		}
+	}
+
+	err = s.billingService.TriggerInvoice(ctx, billing.InvoiceTriggerServiceInput{
+		InvoiceTriggerInput: billing.InvoiceTriggerInput{
+			Invoice:          invoice.InvoiceID(),
+			Trigger:          input.Trigger,
+			ValidationErrors: validationErrors,
+		},
+		AppType:    appentitybase.AppTypeStripe,
+		Capability: appentitybase.CapabilityTypeCollectPayments,
+	})
+	if err != nil {
+		s.logger.ErrorContext(ctx, "failed to trigger invoice failed trigger")
+		return err
+	}
+
+	s.logger.InfoContext(ctx, "invoice state transition handled successfully", "trigger", input.Trigger)
+
+	return nil
+}
+
+func (s *Service) HandleInvoiceSentEvent(ctx context.Context, input appstripeentity.HandleInvoiceSentEventInput) error {
+	if err := input.Validate(); err != nil {
+		return err
+	}
+
+	ctx = context.WithValue(ctx, StripeInvoiceIDAttributeName, input.Invoice.ID)
+
+	invoice, err := s.getInvoiceByStripeID(ctx, input.AppID, input.Invoice.ID)
+	if err != nil {
+		return err
+	}
+
+	if invoice == nil {
+		return nil
+	}
+
+	ctx = context.WithValue(ctx, InvoiceIDAttributeName, invoice.ID)
+	ctx = context.WithValue(ctx, InvoiceStatusAttributeName, invoice.Status)
+
+	return s.billingService.UpdateInvoiceFields(ctx, billing.UpdateInvoiceFieldsInput{
+		Invoice:          invoice.InvoiceID(),
+		SentToCustomerAt: mo.Some(lo.ToPtr(time.Unix(input.SentAt, 0))),
+	})
+}
+
+func stripeErrorToValidationError(stripeErr *stripe.Error) error {
+	if stripeErr == nil {
+		return nil
+	}
+
+	return billing.NewValidationError(string(stripeErr.Code), stripeErr.Msg)
+}
+
+// getInvoiceByStripeID retrieves an invoice by its stripe ID, it returns nil if the invoice is not found (thus not managed by the app)
+func (s *Service) getInvoiceByStripeID(ctx context.Context, appID appentitybase.AppID, stripeInvoiceID string) (*billing.Invoice, error) {
+	invoices, err := s.billingService.ListInvoices(ctx, billing.ListInvoicesInput{
+		Namespaces: []string{appID.Namespace},
+		ExternalIDs: &billing.ListInvoicesExternalIDFilter{
+			Type: billing.InvoicingExternalIDType,
+			IDs:  []string{stripeInvoiceID},
+		},
+		IncludeDeleted: true,
+		Page: pagination.Page{
+			PageNumber: 1,
+			PageSize:   5,
+		},
+		Expand: billing.InvoiceExpand{},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(invoices.Items) == 0 {
```

Comment:
shouldn't we return NotFound and handle it upstream? Like we do with DB

Reply:
I want to ignore any invoices that we don't know of. E.g. if there are invoices generated on the stripe side, let's ignore them for now.